File: 20.valid-parentheses.js
// @leet start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
	let stack = [];
	for (let i = 0; i < s.length; i++) {
		const key = s[i];
		switch (key) {
			case "{":
			case "(":
			case "[":
				stack.push(key);
				break;
			case "}":
				if (stack.pop() !== "{") return false;
				break;
			case ")":
				if (stack.pop() !== "(") return false;
				break;
			case "]":
				if (stack.pop() !== "[") return false;
				break;
			default:
				break;
		}
	}
	if (stack.length) {
		return false;
	}
	return true;
};
// @leet end

File: 20.valid-parentheses.cpp
// @leet start
#include <stack>
#include <string>
using namespace std;
class Solution {
public:
  bool isValid(string s) {
		stack<char>stack;
		for(char ch:s){
			switch (ch) {
				case '(':
				case '[':
				case '{':
				stack.push(ch);
				break;
				case ')':
				case ']':
				case '}':
				if(stack.empty())return false;
					if((')' == ch && stack.top() != '(' ||
							']' == ch && stack.top() != '[' ||
							'}' == ch && stack.top() != '{')){
						return false;
				}else {
					stack.pop();
					break;
				}
			}
		}
		return stack.empty();
  }
};
